src/puppetlabs/puppetdb/query_eng.clj

```diff
@@ -2,59 +2,124 @@
+  [entity]
+  (get-in @entity-fn-idx [entity :munge]))
+
+(defn get-rec
+  [entity]
+  (get-in @entity-fn-idx [entity :rec]))
+
+(defn orderable-columns
+  [query-rec]
+  (if-not query-rec
+    []
+    (let [projections (:projections (query-rec))]
+      (->> (keys projections)
+           (remove #{"value"})
+           (filter #(:queryable? (get projections %)))
+           (map keyword)))))
+
+(defn query->sql
+  "Converts a vector-structured `query` to a corresponding SQL query which will
+   return nodes matching the `query`."
+  [query entity version paging-options]
+  {:pre  [((some-fn nil? sequential?) query)]
+   :post [(map? %)
+          (jdbc/valid-jdbc-query? (:results-query %))
+          (or (not (:count? paging-options))
+              (jdbc/valid-jdbc-query? (:count-query %)))]}
+  (cond
+    (= :aggregate-event-counts entity)
+    (aggregate-event-counts/query->sql version query paging-options)
+
+    (= :event-counts entity)
+    (event-counts/query->sql version query paging-options)
+
+    (and (= :events entity) (:distinct_resources? paging-options))
+    (events/legacy-query->sql false version query paging-options)
+
+    :else
+    (let [query-rec (get-in @entity-fn-idx [entity :rec])
+          columns (orderable-columns query-rec)]
+      (paging/validate-order-by! columns paging-options)
+      (eng/compile-user-query->sql query-rec query paging-options))))
+
+(defn get-munge-fn
   [entity version paging-options url-prefix]
-  (let [[query->sql munge-fn]
-        (case entity
-          :aggregate-event-counts [aggregate-event-counts/query->sql
-                                   aggregate-event-counts/munge-result-rows]
-          :event-counts [event-counts/query->sql
-                         (event-counts/munge-result-rows (first paging-options))]
-          :facts [facts/query->sql facts/munge-result-rows]
-          :fact-contents [fact-contents/query->sql fact-contents/munge-result-rows]
-          :fact-paths [facts/fact-paths-query->sql facts/munge-path-result-rows]
-          :factsets [factsets/query->sql factsets/munge-result-rows]
-          :catalogs [catalogs/query->sql catalogs/munge-result-rows]
-          :nodes [nodes/query->sql (constantly identity)]
-          :environments [environments/query->sql (constantly identity)]
-          :events [events/query->sql events/munge-result-rows]
-          :edges [edges/query->sql edges/munge-result-rows]
-          :reports [reports/query->sql reports/munge-result-rows]
-          :report-metrics [report-data/metrics-query->sql (report-data/munge-result-rows :metrics)]
-          :report-logs [report-data/logs-query->sql (report-data/munge-result-rows :logs)]
-          :resources [resources/query->sql resources/munge-result-rows])]
-    [#(query->sql version % paging-options) (munge-fn version url-prefix)]))
+  (let [munge-fn (get-munge entity)]
+    (case entity
+      :event-counts
+      ((munge-fn (:summarize_by paging-options)) version url-prefix)
```

Comment:
since we're special casing it here anyways, maybe the `:event-counts` munge function should be changed to take three arguments here instead of return a function of two args